core/math/bvh_pair.inc

```diff
@@ -59,4 +59,14 @@ struct ItemPairs {
 
 		return userdata;
 	}
+
+	// experiment : scale the pairing expansion by the number of pairs.
+	// when the number of pairs is high, the density is high and a lower collision margin is better.
+	// when there are few local pairs, a larger margin is more optimal.
+	real_t scale_expansion_margin(real_t p_margin) const {
+		real_t x = real_t(num_pairs) * (1.0 / 9.0);
+		x = MIN(x, 1.0);
+		x = 1.0 - x;
```

Comment:
Is there any risk to end up with a negative `x` with floating point errors?
If yes, could possibly be changed to something like:
```suggestion
		x = MAX(1.0 - x, 0);
```

Reply:
No I don't think so, 1.0 is defined precisely in floating point afaik, and even then I'm not sure a negative expansion margin would be a problem in this case. Should be fine I think. (I do this all the time so it better be lol :grin: )